### ISSUE TITLE
Extend graph merge

### ISSUE THREAD

**Issue by MKaptein** [scailable/sclblonnx#11]
While `so.merge()` provides some initial merge functionality, it would be nice to be able to combine / merge graphs more extensively (see image for an example).

![image009](https://user-images.githubusercontent.com/5011983/121388769-0bf16900-c94c-11eb-8601-5251ed00253f.jpg)

We would need to think about the interface, but it would make the construction of custom graphs by combining subgraphs much easier.

**Comment by MKaptein** [comment]
After a number of useful additions on this issue (see https://github.com/scailable/sclblonnx/pull/19), it seems to make sense to first flesh out in a bit more detail what the desired behavior should be of (additions to) the merge function. Useful ways to proceed seem two-fold:

1. Create (additional) tests for the merge function (in `test/test_merge.py` to start with) such that any new functionality can be done using test driven development.
2. Create some "jargon" to easily discuss possible additions to the merge function.

Let me give 2. a try below (but please do suggest changes / improvements).

## Discussing graph merge variations

It seems that for all merge purposes a graph can be defined by its inputs and outputs (two lists). Assuming we coin "graphs to be merged" subgraphs `sg`, we could write `sg1{[in1, in2, in3, ...], [out1, out2, out3, ...]}` to identify a graph. From that starting point it seems we can provide meaningful descriptions of merge features. For example, I could detail the current functionality generally as:

* Given two subgraphs `sg1{[in1, in2, in3, ...], [out1, out2, out3, ...]}`, `sg2{[in1, in2, in3, ...], [out1, out2, out3, ...]}`, an ordered list of outputs of `sg1`, `[out1, out2, ...]` called `outputs` and an ordered list of inputs of `sg2`, `[in1, in2, ...]` called `inputs`, create a new graph `g` in which `sg1` is deemed the parent of `sg2` and which has inputs `sg1[in] + (sg2[in] - inputs)` and outputs `(sg1[out] - inputs) + sg2[out]`.

This leads to some practical examples:

* Suppose we have a simple subgraph that adds two tensors `x1` and `x2`, and outputs their `sum`. Next, we have a subgraph that checks whether the `input` is less than some threshold and outputs `check`. If we now want to merge the two we thus have:

```
sg1{[x1, x2], [sum]}  # first subgraph
sg2{[input], [check]}. # second subgraph
outputs = [sum]  # the outputs vector of sg1 to be matched with
inputs = [input]. # the inputs vector of sg2

# result:
g{[x1, x2], [check]}
```

Which is the simplest use case the function was designed to handle.

## Next steps.

If the above "jargon" is useful we can propose new functionality using it, such that we are precise whenever suggesting functionality.

## Suggested additions that are clearly useful:

In https://github.com/scailable/sclblonnx/pull/19 we have at least two things that seem useful as is:

1. Being able to (re) name the internal nodes of the graph `g` when doing a merge; this is useful for repeated merges on the same graph etc. Thus, I would suggest accepting the `append_unique_string` proposal.
2. Being less strict on the checking of the match of `inputs` and `outputs`. This is useful whenever one would like to merge more than two subgraphs in sequence, where the intermediate result might not itself be a fully valid graph yet. Actually, it seems that this would allow behavior such as the desired behavior at the start of this issue.

Any suggestions on changing the signature of the function to make it flexible (and intuitive) are very welcome!

**Comment by MKaptein** [comment]
A number of these ideas are implemented in 44c5078 (and thanks to all the discussion in https://github.com/scailable/sclblonnx/pull/19). I will close this issue for now, but please raise additional issues if you find any problems in the new solution using merge(), join(), split(), and concat().
